Update Effective Dart Design Guidelines to Reflect Dart 3.0 Class Modifiers (#6449)
dart-lang/site-www: src/content/effective-dart/design.md

## Classes and mixins

> **Note:** As of Dart 3.0, class modifiers (e.g., `final`, `base`, `interface`) 

Reviewer: We should only include very basic top-level information under this high-level heading. Let's revert this to what was already there, but I would advise adding one bullet with a sentence or two that lets people know that classes can be restricted.

 I believe you've already captured these other details in subsections. If something is here but not in one of those sections, let's move the content to that section.

## Classes and mixins

> Classes. But Dart does not require all code to be defined inside a

Reviewer 2: The word "class" shouldn't be capitalized here.

```suggestion
classes. But Dart does not require all code to be defined inside a
```

class—you can define top-level variables, constants, and functions like

> you can in a procedural or functional language

Reviewer 2: Missing period:

```suggestion
you can in a procedural or functional language.
```

## Classes and mixins

> Classes. But Dart does not require all code to be defined inside a

Reviewer: Original text is still missing.

## Classes and mixins

> Dart is a "pure" object-oriented language in that all objects are instances of

Reviewer: Re-add this.

Otherwise, later changes to it may break your code.

### DO use class modifiers to control if your class can be extended

Class modifiers like `final`, `base`, or `sealed` 

> allow you to enforce whether a class can be extended directly in code. 

Reviewer 2: ```suggestion
enforce whether a class can be extended. 
```

For example, use `final class A {}` to prevent extension, 
or `base class B {}` to allow extension only within the same library. 

> Rely on these modifiers rather than documentation to communicate and 
> enforce your intent.

Reviewer 2: ```suggestion
Use these modifiers to communicate your intent, rather than relying on documentation.
```

> ### ### DO use class modifiers to control if your class can be an interface

Reviewer 2: Your proposed rule is from the perspective of someone writing the library, but the original version was intended for those who are using a library. This section is talking about classs that aren't intended to be an interface, and while it would be nice if all such classes were using class modifiers to enforce these rules (as recommended above), it's still possible that a developer could run into these problems so I would vote to keep this section as-is.

Otherwise, later changes to it may break your code.

### DO document if your class supports being extended
### DO use class modifiers to control if your class can be extended

This is the corollary to the above rule. If you want to allow subclasses of your
class, state that. Suffix the class name with `Base`, or mention it in the
class's doc comment.
Class modifiers like `final`, `base`, or `sealed`
enforce whether a class can be extended.
For example, use `final class A {}` to prevent extension, 
or `base class B {}` to allow extension only within the same library.
Use these modifiers to communicate your intent, rather than relying on documentation.

> ### AVOID implementing a class that isn't intended to be an interface

Author: that was the old heading ig 
it was updated to 

 _### DO use class modifiers to control if your class can be an interface_

> ### AVOID implementing a class that isn't intended to be an interface

Author: just checked my code base 
i guess this was mentioned to be removed and '### DO use class modifiers to control if your class can be an interface'
is suppose to wrap the content of it too

if u need i can add '### AVOID implementing a class that isn't intended to be an interface' and its following contents back in along with '### DO use class modifiers to control if your class can be an interface'

> ### AVOID implementing a class that isn't intended to be an interface

Reviewer 2: Implicit interfaces are still supported in Dart, so I think this section is still useful. I think you're changing the original meaning a bit more than is necessary. For example:

> To give the authors of the classes you use more leeway, avoid implementing
implicit interfaces except for classes that are clearly intended to be
implemented.

This is still true in Dart 3, it's possible that developers will try to implement a class that wasn't intended to be implemented (and is not marked with a class modifier like `final` ).

> ### AVOID implementing a class that isn't intended to be an interface

Author: thanks for your input 
lemme rephrase the para a bit that aligns your POV

> ### AVOID implementing a class that isn't intended to be an interface

Reviewer 2: I think I need to be more clear what I'm asking, thanks for your patience. You are making changes to three rules:

- (Updated) DO use class modifiers to control if your class can be extended
- (Deleted) AVOID implementing a class that isn't intended to be an interface
- (Updated) DO use class modifiers to control if your class can be an interface

I wasn't asking you to rephrase #3, so you can change that back to match #1. I'd prefer to leave this second section as-is for now, and discuss removing it in a separate issue, since it still applies IMO. Thanks!

> ### AVOID implementing a class that isn't intended to be an interface

Author: made the suggested changes 
kindly take a look

Otherwise, later changes to it may break your code.

### DO use class modifiers to control if your class can be extended

Class modifiers like `final`, `interface`, or `sealed`
restrict how a class can be extended.

> For example, use `final class A {}` or `base class B {}` to prevent 

Reviewer 3: It can't be `base class B {}` because that does not prevent extension outside the current library. It's `interface` that prevents `extends` (but allows `implements`).

Author: changed `base class B {}` to 'interface class B{}'

### DO use class modifiers to control if your class can be extended

Class modifiers like `final`, `base`, or `sealed` 
allow you to enforce whether a class can be extended directly in code. 
For example, use `final class A {}` to prevent extension, 
or `base class B {}` to allow extension only within the same library. 
Rely on these modifiers rather than documentation to communicate and 
enforce your intent.

> ### ### DO use class modifiers to control if your class can be an interface

Reviewer: Delete extra ###

> you can also restrict implementation using

Reviewer: You can restrict implementation using

Otherwise, later changes to it may break your code.

### DO use class modifiers to control if your class can be extended

> Since Dart 3.0, class modifiers like `final`, `base`, or `sealed` 

Reviewer: ```suggestion
Class modifiers like `final`, `base`, or `sealed` 
```

Dart 3 has been out for a while. I think we can avoid mentioning it. @johnpryan, your thoughts?

Reviewer 2: I agree - we can leave this out here.

allow you to enforce whether a class can be extended directly in code. 
For example, use `final class A {}` to prevent extension, 
or `base class B {}` to allow extension only within the same library. 
Rely on these modifiers rather than documentation to communicate and 
enforce your intent.

> ### DO use class modifiers to control if your class can be used as an interface

Reviewer: ```suggestion
### DO use class modifiers to control if your class can be an interface
```

I think we can cut a few words from this without removing meaning.

> With Dart 3.0’s class modifiers, you can restrict implementation using

Reviewer: Can you have this section mirror the last one as closely as possible? It's okay if they sound very similar. I like the wording you used in the previous heading for extending.

implemented. Otherwise, you may introduce a coupling that the author doesn't
intend, and they may break your code without realizing it.

### DO use class modifiers to control if your class can be an interface

> When designing a library, use class modifiers like `final`, `base`, or `interface` to enforce intended

Reviewer 3: When it comes to `implements` clauses, `interface` has no effect, but `sealed` does:
```suggestion
When designing a library, use class modifiers like `final`, `base`, or `sealed` to enforce intended
```

> usage. For example, `final class C {}` prevents implementation, while `interface class D {}`

Reviewer 3: ```suggestion
usage. For example, use `final class C {}` or `base class D {}` to prevent
implementation outside the current library.
```

> explicitly allows it. While it's ideal for all libraries to use these modifiers to enforce design intent,

Reviewer 3: ```suggestion
While it's ideal for all libraries to use these modifiers to enforce design intent,
```
